crmd/lrm_state.c

```diff
@@ -286,6 +286,35 @@ lrm_state_get_list(void)
     return g_hash_table_get_values(lrm_state_table);
 }
 
+static void
+remote_proxy_disconnect_by_node(const char * node_name)
+{
+    GListPtr proxies = NULL;
+    GListPtr gIter = NULL;
+
+    CRM_CHECK(proxy_table != NULL, return);
+
+    proxies = g_hash_table_get_values(proxy_table);
+
+    for (gIter = proxies; gIter != NULL; gIter = gIter->next) {
+        remote_proxy_t *proxy = gIter->data;
+
+        if (proxy->source
+            && safe_str_eq(node_name, proxy->node_name)) {
+            /* mainloop_del_ipc_client() eventually calls remote_proxy_disconnected()
+             * , which removes the entry from proxy_table.
+             * Do not do this in a g_hash_table_iter_next() loop. */
```

Comment:
Even this may not be not enough. The documentation for g_hash_table_get_values() says "The returned data is valid until hash_table is modified", which implies this loop could become invalid. In practice it may not, but I don't know if we want to rely on that implementation. Probably the only safe way would be to restart the loop after a hit; that would be inefficient, but this table should generally be small.

Alternatively we could add a flag to the proxy type, set it here, and remote_proxy_disconnected() could skip the removal if the flag is set. Then this whole loop could probably become a g_hash_table_foreach_remove().